### PR TITLE
perf(animations): always run the animation queue outside of zones

### PR DESCRIPTION
**Pull request by matsko**
Related #12732

### CODE REVIEW

File: modules/@angular/compiler/src/view_compiler/constants.ts [hunk 1]
```diff
@@ -42,6 +42,7 @@ export class ViewConstructorVars {
   static parentView = o.variable('parentView');
   static parentIndex = o.variable('parentIndex');
   static parentElement = o.variable('parentElement');
+  static animationQueue = o.variable('animationQueue');
```
Comment (mhevery) [inline modules/@angular/compiler/src/view_compiler/constants.ts]: revert

File: modules/@angular/core/src/animation/animation_queue.ts [hunk 2]
```diff
@@ -5,30 +5,50 @@
...
+    // each request. This sidesteps the issue. If a user
+    // hooks into an animation via (@anim.start) or (@anim.done)
+    // then those methods will automatically trigger change
+    // detection by wrapping themselves inside of a zone
+    this._zone.runOutsideAngular(() => {
+      // this code is wrapped into a single promise such that the
+      // onStart and onDone player callbacks are triggered outside
+      // of the digest cycle of animations
+      if (this.entries.length) {
+        Promise.resolve(null).then(() => this._triggerAnimations());
```
Comment (mhevery) [inline modules/@angular/core/src/animation/animation_queue.ts]: put the `entries.length` check outside the `runOutsideAngular`

File: modules/@angular/core/src/animation/animation_queue.ts [hunk 3]
```diff
@@ -5,30 +5,50 @@
...
+  private _triggerAnimations() {
+    NgZone.assertNotInAngularZone();
+
+    for (let i = 0; i < this.entries.length; i++) {
+      const player = this.entries[i];
+
+      // in the event that an animation throws an error then we do
+      // not want to re-run animations on any previous animations
+      // if they have already been kicked off beforehand
+      if (!player.hasStarted()) {
```
Comment (mhevery) [inline modules/@angular/core/src/animation/animation_queue.ts]: Discussed to `shift()` players from queue.

File: modules/@angular/core/src/animation/animation_transition.ts [hunk 4]
```diff
@@ -8,6 +8,8 @@
 import {AnimationPlayer} from './animation_player';
 import {AnimationTransitionEvent} from './animation_transition_event';
 
+declare let Zone: any;
```
Comment (mhevery) [inline modules/@angular/core/src/animation/animation_transition.ts]: revert

File: modules/@angular/core/src/animation/animation_transition.ts [hunk 5]
```diff
@@ -24,11 +26,13 @@ export class AnimationTransition {
 
   onStart(callback: (event: AnimationTransitionEvent) => any): void {
     const event = this._createEvent('start');
-    this._player.onStart(() => callback(event));
+    const fn = <any>Zone.current.wrap((change: any) => callback(event));
```
Comment (mhevery) [inline modules/@angular/core/src/animation/animation_transition.ts]: remove `<any>` add source argument.

File: modules/@angular/core/src/animation/animation_transition.ts [hunk 6]
```diff
@@ -24,11 +26,13 @@ export class AnimationTransition {
...
+    this._player.onStart(fn);
   }
 
   onDone(callback: (event: AnimationTransitionEvent) => any): void {
     const event = this._createEvent('done');
-    this._player.onDone(() => callback(event));
+    const fn = <any>Zone.current.wrap((change: any) => callback(event));
```
Comment (mhevery) [inline modules/@angular/core/src/animation/animation_transition.ts]: remove `<any>` add source argument.

**Review by mhevery — CHANGES_REQUESTED** [review]

**Review by mhevery — APPROVED** [review]

**Comment by vicb** [comment]
status ?

**Comment by vicb** [comment]
Have the last 2 commits been reviewed ?

**Comment by matsko** [comment]
@vicb Misko has reviewed both commits.

**Pull request closed without merging**